cpp/ql/lib/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll

```diff
@@ -1530,6 +1530,34 @@ private predicate isUnsupportedGuardPhi(Variable v, RangeSsaDefinition phi, Vari
   )
 }
 
+/**
+ * Gets the upper bound of the expression, if the expression is guarded.
+ * An upper bound can only be found, if a guard phi node can be found, and the
+ * expression has only one immediate predecessor.
+ */
+private float getGuardedUpperBound(VariableAccess guardedAccess) {
+  exists(
+    RangeSsaDefinition def, StackVariable v, VariableAccess guardVa, Expr guard, boolean branch
+  |
+    def.isGuardPhi(v, guardVa, guard, branch) and
+    // If the basic block for the variable access being examined has
+    // more than one predecessor, the guard phi node could originate
+    // from one of the predecessors. This is because the guard phi
+    // node is attached to the block at the end of the edge and not on
+    // the actual edge. It is therefore not possible to determine which
+    // edge the guard phi node belongs to. The predicate below ensures
+    // that any predecessor other than the guard block is dominated by the
+    // guard phi node.
+    forall(BasicBlock pred |
+      pred = def.(BasicBlock).getAPredecessor() and pred != guard.getBasicBlock()
+    |
+      bbDominates(def, pred)
+    ) and
```

Comment:
This condition requires `def` to dominate its own predecessor. I can't think of any case where that happens in practice, but I agree with @hvitved that it could in principle happen if we changed the CFG so the back edges of loops would point to the loop statements themselves.

Do the tuple counts still look good?

Reply:
I'll have an extra look to verify that they look sane.

Reply:
After looking at the row counts, I see that `bbDominates` returns a lot of rows and it is not cached for cpp. 
Since `getGuardedUpperBound` it quite limited in its functionality already (it only looks at the immediate guard phi) I have elected to revert to the previous version, as this change only catches quite specific cases.

Reply:
@hvitved

Reply:
I am surprised that `bbIdominates` is not cached. However, `getGuardedUpperBound` is only used in `upperBound`, which is cached, so I would only expect one extra evaluation of `bbIdominates` in a full analysis, so I wouldn't expect any performance regressions. Another reason why I prefer the change is that it aligns the logic with how it is done for other languages (in addition to the C# link above, there is also https://github.com/github/codeql/blob/main/java/ql/lib/semmle/code/java/controlflow/Dominance.qll#L92)

Reply:
We should probably cache `bbIdominates`, but this PR doesn't seem like the right time.

I think we're doing it right in `Guards.qll`:
https://github.com/github/codeql/blob/4c747090191f1847a78125ed0eef4d1b23b0cc89/cpp/ql/lib/semmle/code/cpp/controlflow/Guards.qll#L112-L115

The 1-step improvement of `upperBound` in this PR is supposed to be the simplest and least risky thing we can possibly do. I think the next step is to use `Guards.qll`, and then we'll be deleting this ad hoc criterion about multiple predecessors.